plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/kvm/ha/KvmHaAgentClient.java

```diff
@@ -0,0 +1,295 @@
+    }
+
+    protected int getKvmHaMicroservicePortValue() {
+        Integer haAgentPort = KVMHAConfig.KvmHaWebservicePort.value();
+        if (haAgentPort == null) {
+            LOGGER.warn(String.format("Using default kvm.ha.webservice.port: %s as it was set to NULL for the cluster [id: %d] from %s.",
+                    KVMHAConfig.KvmHaWebservicePort.defaultValue(), agent.getClusterId(), agent));
+            haAgentPort = Integer.parseInt(KVMHAConfig.KvmHaWebservicePort.defaultValue());
+        }
+        return haAgentPort;
+    }
+
+    /**
+     * Checks if the KVM HA Webservice is enabled or not; if disabled then CloudStack ignores HA validation via the webservice.
+     */
+    public boolean isKvmHaWebserviceEnabled() {
+        return KVMHAConfig.IsKvmHaWebserviceEnabled.value();
+    }
+
+    /**
+     * Lists VMs on host according to vm_instance DB table. The states considered for such listing are: 'Running', 'Stopping', 'Migrating'.
+     * <br>
+     * <br>
+     * Note that VMs on state 'Starting' are not common to be at the host, therefore this method does not list them.
+     * However, there is still a probability of a VM in 'Starting' state be already listed on the KVM via '$virsh list',
+     * but that's not likely and thus it is not relevant for this very context.
+     */
+    protected List<VMInstanceVO> listVmsOnHost(Host host, VMInstanceDao vmInstanceDao) {
+        List<VMInstanceVO> listByHostAndStateRunning = vmInstanceDao.listByHostAndState(host.getId(), VirtualMachine.State.Running);
+        List<VMInstanceVO> listByHostAndStateStopping = vmInstanceDao.listByHostAndState(host.getId(), VirtualMachine.State.Stopping);
+        List<VMInstanceVO> listByHostAndStateMigrating = vmInstanceDao.listByHostAndState(host.getId(), VirtualMachine.State.Migrating);
+
+        List<VMInstanceVO> listByHostAndState = new ArrayList<>();
+        listByHostAndState.addAll(listByHostAndStateRunning);
+        listByHostAndState.addAll(listByHostAndStateStopping);
+        listByHostAndState.addAll(listByHostAndStateMigrating);
+
+        if (LOGGER.isTraceEnabled()) {
+            List<VMInstanceVO> listByHostAndStateStarting = vmInstanceDao.listByHostAndState(host.getId(), VirtualMachine.State.Starting);
+            int startingVMs = listByHostAndStateStarting.size();
+            int runningVMs = listByHostAndStateRunning.size();
+            int stoppingVms = listByHostAndStateStopping.size();
+            int migratingVms = listByHostAndStateMigrating.size();
+            int countRunningVmsOnAgent = countRunningVmsOnAgent();
+            LOGGER.trace(
+                    String.format("%s has (%d Starting) %d Running, %d Stopping, %d Migrating. Total listed via DB %d / %d (via libvirt)", agent.getName(), startingVMs, runningVMs,
+                            stoppingVms, migratingVms, listByHostAndState.size(), countRunningVmsOnAgent));
+        }
+
+        return listByHostAndState;
+    }
```

Comment:
As the method `listByHostAndStat` receive a varargs of state as parameter, we could simplify this method by joining the three first requests and filtering they if needed:

```java
List<VMInstanceVO> listByHostAndStates = vmInstanceDao.listByHostAndState(host.getId(), VirtualMachine.State.Running, VirtualMachine.State.Stopping, VirtualMachine.State.Migrating);

if (LOGGER.isTraceEnabled()) {
    List<VMInstanceVO> listByHostAndStateStarting = vmInstanceDao.listByHostAndState(host.getId(), VirtualMachine.State.Starting);
    int startingVMs = listByHostAndStateStarting.size();
    int runningVMs = listByHostAndStates.stream().filter...;
    int stoppingVms = listByHostAndStates.stream().filter...;
    int migratingVms = listByHostAndStates.stream().filter...;
    int countRunningVmsOnAgent = countRunningVmsOnAgent();
    LOGGER.trace(
            String.format("%s has (%d Starting) %d Running, %d Stopping, %d Migrating. Total listed via DB %d / %d (via libvirt)", agent.getName(), startingVMs, runningVMs,
                    stoppingVms, migratingVms, listByHostAndState.size(), countRunningVmsOnAgent));
}

return listByHostAndStates;
```